README.md

```diff
@@ -59,8 +59,23 @@ Once you have implemented a lower bound search for the start of the range, imple
 *Answer the questions below and then push this file to GitHub.*
 
 *How do you use binary search to find the lower bound of a number? How did you have to modify the binary search algorithm?*
+The classical binary search is checking to see if a specidic number is in a list and is coded like: 
+low, high = 0, len(numbers)
+found = False
+while low < high:
+    mid = (low + high) // 2
+    if numbers[mid] == x:
+        found = True
+        break
+    elif numbers[mid] < x:
+        low = mid + 1
+    else:
+        high = mid
+I am not looking to see if a specific number is in a list, I am looking to see what index a specific number is at in my list, so I can remember that index site and use it to give me the region I am interested in. I had to modify the binary search by having the while loop not break when it finds the correct value like it does in classic binary search and return a boolean vale of true if it is there or false if it is not, but instead break when start >= end, ie when I can return start which will be the index number for where my integer that I am looking for can be found. 
 
 *Would anything be more difficult if the features covered ranges instead of single nucleotides (like real BED files)? What could go wrong, if anything?*
+If features covered ranges instead of single nucleotides,  my sort_chromosome function would still be able to sort it without more difficulty (I think). Right now, it sorts first on chromosome start, then by chromosome end then by name in that order of importance because I used key= attrgetter for my sort operator. I think right now, if this function were given ranges instead of single nucleotides, it would order them by their start index first, and then if two regions had the same start index, the range with the first end index would be sorted first, and then if they had the same start and end region, the range with the first name alphabetically would go first. 
 
 *We wrote a tool for merging two BED files, but what if we had a bunch of them? What would the complexity be if we merged them in, one at a time? What would the complexity be if we merged all of the files at the same time?*
 
+I we merged them in one at a time, the complexity would be O(n + m (Sum(len(other functions))) because we would have to through all of the functions every time we added a new one. If we merged them at the same time, I think the complexity would be O Σ(len(functions)) because I would only have to go through all of the functions once. 
```

Comment:
I'm not entirely sure that I understand the first formula. It might be right. There is certainly a sum of increasing lengths.

For simplicity, let’s say that all the files have length $m$ and that there are $n$ of them. Merging the first two would take time $2m$ (two files of length $m$). Merging in the third would take $3m$, however (one file of length $2m$ and one of length $m$). The fourth would require time $4m$ (merging one file of length $3m$ with one of length $m$).

In general, we would need the time $m(2+3+4+\cdots+n-1)$ time to merge $n$ files of length $m$ and the sum in that expression (as you know) is in $O(n^2)$, so the running time is $O(mn^2)$ (far from linear).

The second result is a bit optimistic. If you only output each element, then you would get $O(nm)$ (your formula with the variables I just used), because you only need to read $nm$ elements and output them again, but you need to _find_ the smallest element out of $n$ in each iteration, and if you use a linear scan to find the minimum you are back to $O(n^2m)$.

Concatenating all the files and sorting would instead take $O(nm\log(nm))$, which could be better (depending on $n$ and $m$).


You could also, using divide and conquer, split the input in two $n/2$-length lists of files, merge these recursively, and then merge the results. That would give you the recurrence $T(n,m) = 2T(n/2,m) + O(n\cdot m)$ which is in $O(mn\log(n))$.

If you merge the $n$ files in parallel, so you, in each iteration, find the smallest front element in the $n$ files, you need to copy $nm$ elements, but it takes time to find the smallest. If you do this with a linear scan, it takes $O(n)$ to find the smallest element in each iteration, and you end up back at $O(mn^2)$. You can speed this up (once we have seen heaps in a later lecture) so it takes time $O(\log n)$ to get the smallest element per iteration, and then you are back to $O(mn\log n)$ (just with a completely different approach).